### PR TITLE
Api/api spec restdocs api spec

### LINKED ISSUE #8: api: add api spec to the project
In order to better control and organize our API specification, we should have it documented and maintained, even if the endpoints have not yet been created. The best option is probably OpenAPI but feel free to try something else if you feel like it'd fit best here (a docusaurus documentation could be nice but it's a lot more work and probably overkill)

### PR DESCRIPTION
**Pull request by Sirze01**
Closes #8

Merging this PR will add the programmatic documentation implementation using restdocs-api-spec, as well as the documentation for currently existing endpoints.

# Review checklist
-   [x] Properly documents API changes in `docs/openapi.yml`
-   [x] Contains enough appropriate tests
-   [x] Behavior is as expected
-   [x] Clean, well structured code

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -77,3 +77,50 @@ jobs:
...
+      NETLIFY_SITE_ID: ${{ vars.NETLIFY_SITE_ID }}
+    steps:
+      - name: Fetch sources
+        uses: actions/checkout@v3
+
+      - name: Setup Java
+        uses: actions/setup-java@v3
+        with:
+          java-version: ${{ env.JAVA_VERSION }}
+          distribution: ${{ env.JAVA_DISTRIBUTION }}
```
Comment (ttoino) [inline .github/workflows/ci.yml:99]: This job should use the `gradle-build-action` like the other ones, thus ensuring that the gradle cache is used.

Comment (Sirze01) [reply .github/workflows/ci.yml:99]: Nice catch... I was wondering if this setup step was still required when using the repo's wrapper. I've read the action documentation and now and it makes sense. Thanks.

Comment (ttoino) [reply .github/workflows/ci.yml:99]: It's technically not required as the `setup-java` action _can_ do it as well, but it's disabled by default and not as powerful, as we cannot set an explicit gradle version like we do right now

**Review by ttoino — APPROVED** [review]
Looks good! Would be cool if the json could be prettyfied in the docs website but it looks really good as is

**Comment by Sirze01** [comment]
> Looks good! Would be cool if the json could be prettified in the docs website but it looks really good as is

It seems the problem is with the plugin I'm using. It considers the snippets generated by Spring Restdocs as a string and not as a JSON object and then redoc doesn't prettify the string.

I think I'll open an issue on their side, but I think it is ok for our use case for now. 

PS: Here is the issue https://github.com/ePages-de/restdocs-api-spec/issues/230

**Comment by BrunoRosendo** [comment]
> > Looks good! Would be cool if the json could be prettified in the docs website but it looks really good as is
> 
> It seems the problem is with the plugin I'm using. It considers the snippets generated by Spring Restdocs as a string and not as a JSON object and then redoc doesn't prettify the string.
> 
> I think I'll open an issue on their side, but I think it is ok for our use case for now.
> 
> PS: Here is the issue [ePages-de/restdocs-api-spec#230](https://github.com/ePages-de/restdocs-api-spec/issues/230)

I see they closed it and posted a workaround, should we open an issue in our end to use that?

File: README.md [hunk 2]
```diff
@@ -66,6 +66,28 @@ Run the following command in your shell:
...
+##### With the command line
+Run the following command in your shell:
+
+```bash
+gradle openapi3
+```
+
+```bash
+gradle postman
+```
```
Comment (BrunoRosendo) [inline README.md]:
```suggestion
```bash
./gradlew openapi3
```

```bash
./gradlew postman
```
```

wdyt?

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/AccountControllerTest.kt [hunk 3]
```diff
@@ -22,14 +30,24 @@ import pt.up.fe.ni.website.backend.utils.TestUtils
...
+import pt.up.fe.ni.website.backend.utils.documentation.DocumentedJSONField
+import pt.up.fe.ni.website.backend.utils.documentation.MockMVCExtension.Companion.andDocument
+import pt.up.fe.ni.website.backend.utils.documentation.MockMVCExtension.Companion.andDocumentCustomRequestSchemaEmptyResponse
+import pt.up.fe.ni.website.backend.utils.documentation.MockMVCExtension.Companion.andDocumentCustomRequestSchemaErrorResponse
+import pt.up.fe.ni.website.backend.utils.documentation.MockMVCExtension.Companion.andDocumentErrorResponse
+import pt.up.fe.ni.website.backend.utils.documentation.ModelDocumentation
+import pt.up.fe.ni.website.backend.utils.documentation.PayloadSchema
 
 @ControllerTest
+@AutoConfigureRestDocs
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/AccountControllerTest.kt]: What do you think about including this in `@ControllerTest`?

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/AccountControllerTest.kt [hunk 4]
```diff
@@ -88,33 +112,50 @@ class AccountControllerTest @Autowired constructor(
             repository.save(testAccount)
         }
 
+        private val parameters = listOf(
+            parameterWithName("id").description(
+                "ID of the account to retrieve"
+            )
+        )
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/AccountControllerTest.kt:117]: I have a question. Is this for path and query parameters? What about the body request (I'm guessing it's inferred)?

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/AccountControllerTest.kt [hunk 5]
```diff
@@ -128,23 +169,32 @@ class AccountControllerTest @Autowired constructor(
...
+                    jsonPath("$.github").value(testAccount.github),
+                    jsonPath("$.websites.length()").value(1),
+                    jsonPath("$.websites[0].url").value(testAccount.websites[0].url),
+                    jsonPath("$.websites[0].iconPath").value(testAccount.websites[0].iconPath)
+
+                )
+                .andDocument(
+                    documentation,
+                    "Create new accounts",
+                    "This endpoint operation creates new accounts.",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/AccountControllerTest.kt]:
```suggestion
                    "This endpoint operation creates a new account.",
```

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/AccountControllerTest.kt [hunk 6]
```diff
@@ -435,22 +499,47 @@ class AccountControllerTest @Autowired constructor(
...
+            )
+        )
+
+        private val passwordChangePayload = PayloadSchema(
+            "password-change",
+            mutableListOf(
+                DocumentedJSONField("oldPassword", "Current account password", JsonFieldType.STRING),
+                DocumentedJSONField("newPassword", "New account password", JsonFieldType.STRING)
+            )
+        )
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/AccountControllerTest.kt:512]: Why do we need this here and not when creating an account for example? Got me a bit confused

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/AuthControllerTest.kt [hunk 7]
```diff
@@ -54,43 +72,62 @@ class AuthControllerTest @Autowired constructor(
...
+                .andExpectAll(
+                    status().isOk,
+                    jsonPath("$.access_token").exists(),
+                    jsonPath("$.refresh_token").exists()
+                )
+                .andDocument(
+                    documentation,
+                    "Authenticate account",
+                    "This endpoint operation allows authentication using user's password and email, " +
+                        "generating new access and refresh tokens to be used in later communication.",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/AuthControllerTest.kt]:
```suggestion
                        "generating new access and refresh tokens to be used in following requests.",
```
wdyt?

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/AuthControllerTest.kt [hunk 8]
```diff
@@ -120,13 +162,22 @@ class AuthControllerTest @Autowired constructor(
...
+                        .content(objectMapper.writeValueAsString(TokenDto(refreshToken)))
+                )
+                    .andExpectAll(
+                        status().isOk,
+                        jsonPath("$.access_token").exists()
+                    )
+                    .andDocument(
+                        documentation,
+                        "Refresh access token",
+                        "This endpoint operation allows the renewal of access tokens when expiring, " +
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/AuthControllerTest.kt]: I think it works even when the token is not expired, but please @bdmendes confirm

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt [hunk 9]
```diff
@@ -93,12 +106,16 @@ internal class EventControllerTest @Autowired constructor(
...
-                }
+            mockMvc.perform(get("/events").accept(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk)
+                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
+                .andExpect(content().json(objectMapper.writeValueAsString(testEvents)))
+                .andDocument(
+                    documentation.getModelDocumentationArray(),
+                    "Get all the events",
+                    "The operation returns an array of events, allowing to easily retrieve all the created events.\n" +
+                        "This is useful for example in the frontend's event page, where events are displayed."
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt]: Idk if it's worth it here but with multi-line string there'd be no need of \n I think: https://kotlinlang.org/docs/java-to-kotlin-idioms-strings.html#use-multiline-strings

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt [hunk 10]
```diff
@@ -257,37 +309,46 @@ internal class EventControllerTest @Autowired constructor(
...
+                    jsonPath("$.dateInterval.endDate").value(testEvent.dateInterval.endDate.toJson()),
+                    jsonPath("$.location").value(testEvent.location),
+                    jsonPath("$.category").value(testEvent.category),
+                    jsonPath("$.thumbnailPath").value(testEvent.thumbnailPath),
+                    jsonPath("$.slug").value(testEvent.slug)
+                )
+                .andDocument(
+                    documentation,
+                    "Create new events",
+                    "This endpoint operation creates new events.",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt]:
```suggestion
                    "This endpoint operation creates a new event.",
```

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt [hunk 11]
```diff
@@ -539,46 +615,59 @@ internal class EventControllerTest @Autowired constructor(
...
+                    jsonPath("$.teamMembers[1].linkedin").value(newAccount.linkedin),
+                    jsonPath("$.teamMembers[1].github").value(newAccount.github),
+                    jsonPath("$.teamMembers[1].websites.length()").value(1),
+                    jsonPath("$.teamMembers[1].websites[0].url").value(newAccount.websites[0].url),
+                    jsonPath("$.teamMembers[1].websites[0].iconPath").value(newAccount.websites[0].iconPath)
+                )
+                .andDocument(
+                    documentation,
+                    "Add member to event",
+                    "This operation adds a member to a given event.",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt]:
```suggestion
                    "This operation adds a team member to a given event.",
```

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt [hunk 12]
```diff
@@ -592,25 +681,38 @@ internal class EventControllerTest @Autowired constructor(
...
-                }
+            mockMvc.perform(put("/events/{eventId}/removeTeamMember/{accountId}", testEvent.id, testAccount.id))
+                .andExpectAll(
+                    status().isOk,
+                    content().contentType(MediaType.APPLICATION_JSON),
+                    jsonPath("$.teamMembers.length()").value(0)
+                )
+                .andDocument(
+                    documentation,
+                    "Remove member from event",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt]:
```suggestion
                    "Remove team member from event",
```

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt [hunk 13]
```diff
@@ -592,25 +681,38 @@ internal class EventControllerTest @Autowired constructor(
...
+                    "This operation removes a member of a given event.",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt]:
```suggestion
                    "This operation removes a team member of a given event.",
```

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt [hunk 14]
```diff
@@ -539,46 +615,59 @@ internal class EventControllerTest @Autowired constructor(
...
+                    jsonPath("$.teamMembers[1].photoPath").value(newAccount.photoPath),
+                    jsonPath("$.teamMembers[1].linkedin").value(newAccount.linkedin),
+                    jsonPath("$.teamMembers[1].github").value(newAccount.github),
+                    jsonPath("$.teamMembers[1].websites.length()").value(1),
+                    jsonPath("$.teamMembers[1].websites[0].url").value(newAccount.websites[0].url),
+                    jsonPath("$.teamMembers[1].websites[0].iconPath").value(newAccount.websites[0].iconPath)
+                )
+                .andDocument(
+                    documentation,
+                    "Add member to event",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt]:
```suggestion
                    "Add team member to event",
```

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt [hunk 15]
```diff
@@ -636,35 +740,49 @@ internal class EventControllerTest @Autowired constructor(
...
+                    jsonPath("$.location").value(newLocation),
+                    jsonPath("$.category").value(newCategory),
+                    jsonPath("$.thumbnailPath").value(newThumbnailPath),
+                    jsonPath("$.slug").value(newSlug)
+
+                )
+                .andDocument(
+                    documentation,
+                    "Update events",
+                    "Update previously created events, using their ID.",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt]:
```suggestion
                    "Update a previously created event, using its ID.",
```

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/PostControllerTest.kt [hunk 16]
```diff
@@ -140,20 +198,27 @@ internal class PostControllerTest @Autowired constructor(
...
+                    jsonPath("$.body").value(testPost.body),
+                    jsonPath("$.thumbnailPath").value(testPost.thumbnailPath),
+                    jsonPath("$.publishDate").exists(),
+                    jsonPath("$.lastUpdatedAt").exists(),
+                    jsonPath("$.slug").value(testPost.slug)
+                )
+                .andDocument(
+                    documentation,
+                    "Create new posts",
+                    "This endpoint operation creates new posts.",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/PostControllerTest.kt]:
```suggestion
                    "This endpoint operation creates a new post.",
```

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/ProjectControllerTest.kt [hunk 17]
```diff
@@ -163,31 +219,40 @@ internal class ProjectControllerTest @Autowired constructor(
...
+                    jsonPath("$.teamMembers[0].email").value(testAccount.email),
+                    jsonPath("$.teamMembers[0].name").value(testAccount.name),
+                    jsonPath("$.technologies.length()").value(testProject.technologies.size),
+                    jsonPath("$.technologies[0]").value(testProject.technologies[0]),
+                    jsonPath("$.slug").value(testProject.slug)
+                )
+                .andDocument(
+                    documentation,
+                    "Create new projects",
+                    "This endpoint operation creates new projects.",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/ProjectControllerTest.kt]:
```suggestion
                    "This endpoint operation creates a new project.",
```

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/ProjectControllerTest.kt [hunk 18]
```diff
@@ -326,32 +409,44 @@ internal class ProjectControllerTest @Autowired constructor(
...
+                    content().contentType(MediaType.APPLICATION_JSON),
+                    jsonPath("$.title").value(newTitle),
+                    jsonPath("$.description").value(newDescription),
+                    jsonPath("$.teamMembers.length()").value(0),
+                    jsonPath("$.isArchived").value(newIsArchived)
+                )
+                .andDocument(
+                    documentation,
+                    "Update projects",
+                    "Update previously created projects, using their ID.",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/ProjectControllerTest.kt]:
```suggestion
                    "Update a previously created project, using its ID.",
```

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/ProjectControllerTest.kt [hunk 19]
```diff
@@ -528,16 +653,30 @@ internal class ProjectControllerTest @Autowired constructor(
...
+                .andExpectAll(
+                    status().isOk,
+                    content().contentType(MediaType.APPLICATION_JSON),
+                    jsonPath("$.isArchived").value(newIsArchived)
+                )
+                .andDocument(
+                    documentation,
+                    "Archive projects",
+                    "This endpoint updates projects as archived. This is useful to mark no longer " +
+                        "maintained or complete projects of the Nucleus.",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/ProjectControllerTest.kt]:
```suggestion
                    "This endpoint archives a project. This is useful to mark no longer " +
                        "maintained or complete projects.",
```
Nucleus seems irrelevant for the API

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/ProjectControllerTest.kt [hunk 20]
```diff
@@ -561,16 +700,27 @@ internal class ProjectControllerTest @Autowired constructor(
...
+            )
+                .andExpectAll(
+                    status().isOk,
+                    content().contentType(MediaType.APPLICATION_JSON),
+                    jsonPath("$.isArchived").value(newIsArchived)
+                )
+                .andDocument(
+                    documentation,
+                    "Unarchive projects",
+                    "This endpoint updates projects as unarchived. " +
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/ProjectControllerTest.kt]: same as above

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/ProjectControllerTest.kt [hunk 21]
```diff
@@ -602,44 +752,63 @@ internal class ProjectControllerTest @Autowired constructor(
...
+                    jsonPath("$.teamMembers[1].linkedin").value(newAccount.linkedin),
+                    jsonPath("$.teamMembers[1].github").value(newAccount.github),
+                    jsonPath("$.teamMembers[1].websites.length()").value(1),
+                    jsonPath("$.teamMembers[1].websites[0].url").value(newAccount.websites[0].url),
+                    jsonPath("$.teamMembers[1].websites[0].iconPath").value(newAccount.websites[0].iconPath)
+                )
+                .andDocument(
+                    documentation,
+                    "Add member to Project",
+                    "This operation add a member to a given project.",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/ProjectControllerTest.kt]:
```suggestion
                    "Add member to Project",
                    "This operation add a member to a given project.",
```
```suggestion
                    "Add team member to Project",
                    "This operation add a team member to a given project.",
```

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/ProjectControllerTest.kt [hunk 22]
```diff
@@ -653,24 +822,42 @@ internal class ProjectControllerTest @Autowired constructor(
...
+            mockMvc.perform(put("/projects/{projectId}/removeTeamMember/{accountId}", testProject.id, testAccount.id))
+                .andExpectAll(
+                    status().isOk,
+                    content().contentType(MediaType.APPLICATION_JSON),
+                    jsonPath("$.teamMembers.length()").value(0)
+                )
+                .andDocument(
+                    documentation,
+                    "Remove member from Project",
+                    "This operation removes a member of a given project.",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/ProjectControllerTest.kt]:
```suggestion
                    "Remove team member from Project",
                    "This operation removes a team member from a given project.",
```

File: src/test/kotlin/pt/up/fe/ni/website/backend/documentation/payloadschemas/model/Account.kt [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+    Companion.Tag.ACCOUNT.name.lowercase(),
+    Companion.Tag.ACCOUNT,
+    mutableListOf(
+        DocumentedJSONField("name", "Name of the account owner", JsonFieldType.STRING),
+        DocumentedJSONField("email", "Email associated to the account", JsonFieldType.STRING),
+        DocumentedJSONField("bio", "Short profile description", JsonFieldType.STRING, optional = true),
+        DocumentedJSONField("birthDate", "Birth date of the owner", JsonFieldType.STRING, optional = true),
+        DocumentedJSONField(
+            "photoPath",
+            "Path to the photo resource on the backend server",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/documentation/payloadschemas/model/Account.kt]:
```suggestion
            "Path to the photo resource",
```
Often not true, probably uses another cloud. I think this is enough

File: src/test/kotlin/pt/up/fe/ni/website/backend/documentation/payloadschemas/model/Account.kt [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+            JsonFieldType.STRING,
+            optional = true
+        ),
+        DocumentedJSONField(
+            "linkedin",
+            "Handle/link to the owner's LinkedIn profile",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/documentation/payloadschemas/model/Account.kt]: Wdym by handle?

File: src/test/kotlin/pt/up/fe/ni/website/backend/documentation/payloadschemas/model/Post.kt [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+        DocumentedJSONField(
+            "slug",
+            "Short and friendly textual post identifier",
+            JsonFieldType.STRING,
+            optional = true
+        ),
+        DocumentedJSONField("id", "Post ID", JsonFieldType.NUMBER, isInRequest = false),
+        DocumentedJSONField(
+            "publishDate",
+            "Date of publication of the post",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/documentation/payloadschemas/model/Post.kt]:
```suggestion
            "Post's publish date",
```

File: src/test/kotlin/pt/up/fe/ni/website/backend/documentation/payloadschemas/model/Post.kt [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+            JsonFieldType.STRING,
+            isInRequest = false
+        ),
+        DocumentedJSONField(
+            "lastUpdatedAt",
+            "Date of the last update of the post",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/documentation/payloadschemas/model/Post.kt]:
```suggestion
            "Post's latest update",
```
better flow imo

File: src/test/kotlin/pt/up/fe/ni/website/backend/documentation/payloadschemas/model/Project.kt [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+import pt.up.fe.ni.website.backend.utils.documentation.DocumentedJSONField.Companion.addFieldsBeneathPath
+import pt.up.fe.ni.website.backend.utils.documentation.ModelDocumentation
+
+class Project : ModelDocumentation(
+    Companion.Tag.PROJECT.name.lowercase(),
+    Companion.Tag.PROJECT,
+    mutableListOf(
+        DocumentedJSONField("title", "Project title", JsonFieldType.STRING),
+        DocumentedJSONField("description", "Project description", JsonFieldType.STRING),
+        DocumentedJSONField("isArchived", "Whether this project is being actively maintained", JsonFieldType.BOOLEAN),
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/documentation/payloadschemas/model/Project.kt]: the description is telling the opposite of the field

File: src/test/kotlin/pt/up/fe/ni/website/backend/documentation/payloadschemas/model/Account.kt [hunk 28]
```diff
@@ -0,0 +1,52 @@
+package pt.up.fe.ni.website.backend.documentation.payloadschemas.model
+
+import org.springframework.restdocs.payload.JsonFieldType
+import pt.up.fe.ni.website.backend.utils.documentation.DocumentedJSONField
+import pt.up.fe.ni.website.backend.utils.documentation.ModelDocumentation
+
+class Account : ModelDocumentation(
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/documentation/payloadschemas/model/Account.kt]: Do you not like `PayloadAccount` here to avoid changing import name?

**Review by BrunoRosendo — CHANGES_REQUESTED** [review]
Just as discussed, I like the implementation! I'm not a technical writer but I tried to leave suggestions about the documentation itself.

Other than that, I think having `documentation/` and `utils/documentation/` is a bit odd. Wdyt about moving those files to `documentation/`?

File: README.md (hunk 2 above)
Comment (Sirze01) [reply README.md]: Yes... Makes sense. I just copied the examples above it. They have since been updated. I will also update these.

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/AccountControllerTest.kt (hunk 3 above)
Comment (Sirze01) [reply src/test/kotlin/pt/up/fe/ni/website/backend/controller/AccountControllerTest.kt]: I forgot to do it, nice catch.

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/AccountControllerTest.kt (hunk 4 above)
Comment (Sirze01) [reply src/test/kotlin/pt/up/fe/ni/website/backend/controller/AccountControllerTest.kt:117]: The `parameterWithName` method generates a `ParameterDescriptorWithType` object that can be used for both URL and query parameters. I haven't added the option to document query parameters with the extension methods yet as currently, we have no query parameters in any of our endpoints, but I can add it now if you see fit.

Regarding the body request, it is declared in the `DocumentedJSONField` object, which declares whether it a field is present in the request, response or both. At the time of documenting,  a `FieldDescriptor` object (from the restdocs-api-spec package) is generated based on the declared options.  

The next step in granularity would be to make the `optional` and `ignored` options sensitive to whether the field is documenting a request or response, using 4 options instead of only 2 (`optionalInRequest`, `ignoredInRequest`, `optionalInResponse`, `ignoredInResponse`), but the current degree of granularity is fine for now.

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/AccountControllerTest.kt (hunk 6 above)
Comment (Sirze01) [reply src/test/kotlin/pt/up/fe/ni/website/backend/controller/AccountControllerTest.kt:512]: When documenting the account the `PayloadSchema` is inferred from the `Account` `ModelDocumentation`. Here the schema of the request is different as it only sends in the body 
```json
{"oldPassword": "password", "newPassword": "test_password2"}
```
which is not a regular account schema. So the "custom request" documentation methods need to be used.

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/AuthControllerTest.kt (hunk 8 above)
Comment (Sirze01) [reply src/test/kotlin/pt/up/fe/ni/website/backend/controller/AuthControllerTest.kt]: It does. It's just a poor choice of words.

File: src/test/kotlin/pt/up/fe/ni/website/backend/documentation/payloadschemas/model/Account.kt (hunk 24 above)
Comment (Sirze01) [reply src/test/kotlin/pt/up/fe/ni/website/backend/documentation/payloadschemas/model/Account.kt]: I mean the GitHub @, as in @Sirze01. LinkedIn also supports these mentions and "handles", maybe I should change it to the username.

**Comment by Sirze01** [comment]
> Other than that, I think having `documentation/` and `utils/documentation/` is a bit odd. Wdyt about moving those files to `documentation/`?

My thought process was to try and separate what is documentation from the documentation helpers and wrappers. That's why I separated documentation-related files into 2 folders. `documentation/` for the documentation itself and `utils/documentation/` for the documentation helpers and wrappers.

**Comment by BrunoRosendo** [comment]
> 

I understand. It's not that big of a deal but maybe having another utils under docs? Just to separate things

File: src/test/kotlin/pt/up/fe/ni/website/backend/documentation/payloadschemas/model/Account.kt (hunk 24 above)
Comment (BrunoRosendo) [reply src/test/kotlin/pt/up/fe/ni/website/backend/documentation/payloadschemas/model/Account.kt]: Doesn't the field need to be an URL?

**Review by BrunoRosendo — APPROVED** [review]
Great change! Just needs a rebase
#88 has since been merge but, if you want, just create an issue and I'll do it when I can

**Comment by Sirze01** [comment]
> Great change! Just needs a rebase #88 has since been merge but, if you want, just create an issue and I'll do it when I can

I'm just finishing the documentation :).

**Comment by Sirze01** [comment]
Can you check the generation documentation? Tell me if anything needs fixing. @BrunoRosendo

**Comment by Sirze01** [comment]
> > 
> 
> I understand. It's not that big of a deal but maybe having another utils under docs? Just to separate things

Like `/utils/documentation/utils` ? I guess it achieves what I was trying to do.

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/GenerationControllerTest.kt [hunk 29]
```diff
@@ -328,15 +397,22 @@ class GenerationControllerTest @Autowired constructor(
...
+            )
+                .andExpectAll(
+                    status().isOk,
+                    content().contentType(MediaType.APPLICATION_JSON),
+                    jsonPath("$.schoolYear").value("22-23")
+                )
+                .andDocument(
+                    documentation,
+                    "Create new generations",
+                    "This operation retrieves creates a new generation.",
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/GenerationControllerTest.kt]:
```suggestion
                    "This operation creates a new generation.",
```
I think it was there by mistake

**Review by BrunoRosendo — APPROVED** [review]
